File: src/main/java/me/conorthedev/mediamod/util/PlayerMessager.java
package me.conorthedev.mediamod.util;

import net.minecraft.client.Minecraft;
import net.minecraft.client.entity.EntityPlayerSP;
import net.minecraft.util.text.ITextComponent;
import net.minecraft.util.text.TextComponentString;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import net.minecraftforge.fml.common.gameevent.TickEvent;

import java.util.concurrent.ConcurrentLinkedQueue;

public class PlayerMessager {

    public static final PlayerMessager INSTANCE = new PlayerMessager();
    private static final ConcurrentLinkedQueue<String> queuedMessages = new ConcurrentLinkedQueue<>();
    private static final ConcurrentLinkedQueue<ITextComponent> messages = new ConcurrentLinkedQueue<>();

    private PlayerMessager() {
        TickScheduler.INSTANCE.schedule(0, this::check);
    }

    public void queue(String chat) {
        queuedMessages.add(chat);
    }

    public static void sendMessage(ITextComponent message) {
        if (message == null) message = new TextComponentString("");
        messages.add(message);
    }

    public static void sendMessage(String message, boolean header) {
        if (message == null) return;
        if (header) {
            Minecraft.getMinecraft().player.sendMessage(new TextComponentString(ChatColor.translateAlternateColorCodes('&',
                    "&c[&fMediaMod&c]&r " + message)));
        } else {
            Minecraft.getMinecraft().player.sendMessage(new TextComponentString(ChatColor.translateAlternateColorCodes('&', message)));
        }
    }

    public static void sendMessage(String message) {
        if (message == null) return;
        sendMessage(ChatColor.translateAlternateColorCodes('&', message), true);
    }

    private void check() {
        if (!queuedMessages.isEmpty()) {
            EntityPlayerSP player = Minecraft.getMinecraft().player;
            if (player != null && queuedMessages.peek() != null) {
                String poll = queuedMessages.poll();
                sendMessage(poll);
            }
        }
    }

    @SubscribeEvent
    public void tick(TickEvent.ClientTickEvent event) {
        if (Minecraft.getMinecraft().player == null) {
            return;
        }

        while (!messages.isEmpty()) {
            Minecraft.getMinecraft().player.sendMessage(messages.poll());
        }

        while (!queuedMessages.isEmpty()) {
            sendMessage(queuedMessages.poll());
        }
    }
}

File: src/main/kotlin/dev/cbyrne/mediamod/utils/PlayerMessager.kt
package dev.cbyrne.mediamod.utils

import club.sk1er.mods.core.universal.ChatColor
import club.sk1er.mods.core.universal.ChatColor.COLOR_CHAR
import net.minecraft.client.Minecraft
import net.minecraft.client.entity.EntityPlayerSP
import net.minecraft.util.ChatComponentText
import net.minecraft.util.IChatComponent
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
import net.minecraftforge.fml.common.gameevent.TickEvent
import java.util.concurrent.ConcurrentLinkedQueue

object PlayerMessager {
    val queuedMessages: ConcurrentLinkedQueue<String> = ConcurrentLinkedQueue()
    val messages: ConcurrentLinkedQueue<IChatComponent> = ConcurrentLinkedQueue()

    init {
        // todo tick schedule
    }

    private fun check() {
        if(!queuedMessages.isEmpty()) {
            val player: EntityPlayerSP = Minecraft.getMinecraft().thePlayer

            if(player != null && queuedMessages.peek() != null) {
                val poll: String = queuedMessages.poll()
                sendMessage(poll)
            }
        }
    }

    fun sendMessage(message: IChatComponent) {

    }

    fun sendMessage(message: String) {
        message.takeIf { it.isEmpty() }?.let { return }
        sendMessage(ChatColor.translateAlternateColorCodes('&', message), true)
    }

    fun sendMessage(message: String, header: Boolean) {
        message.takeIf { it.isEmpty() }?.let { return }
        if(header) {
            Minecraft.getMinecraft().thePlayer.addChatComponentMessage(
                ChatComponentText(
                    ChatColor.translateAlternateColorCodes('&',
                "&c[&fMediaMod&c]&r $message"
            ))
            )
        } else {
            Minecraft.getMinecraft().thePlayer.addChatComponentMessage(ChatComponentText(ChatColor.translateAlternateColorCodes('&', message)))
        }
    }

    @SubscribeEvent
    fun onTick (e: TickEvent.ClientTickEvent) {
        if(Minecraft.getMinecraft().thePlayer == null) {
            return;
        }

        while(!messages.isEmpty()) {
            Minecraft.getMinecraft().thePlayer.addChatComponentMessage(messages.poll())
        }

        while(!queuedMessages.isEmpty()) {
            sendMessage(queuedMessages.poll())
        }
    }
}